### PR TITLE
Cloud Provider Interface for Kubemark

### PR DESCRIPTION
**Pull request by bskiba**
This allows to run Custer Autoscaler on Kubemark.
See autoscaler/cluster-autoscaler/proposals/kubemark_integration.md
for more details.

### CODE REVIEW

**Comment by mwielgus** [comment]
```
# k8s.io/autoscaler/cluster-autoscaler/core
core/autoscaler_test.go:70: not enough arguments in call to NewAutoscaler
	have (AutoscalerOptions, *simulator.PredicateChecker, *"k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/client-go/kubernetes/fake".Clientset, record.EventRecorder, "k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes".ListerRegistry)
	want (AutoscalerOptions, *simulator.PredicateChecker, "k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/client-go/kubernetes".Interface, record.EventRecorder, "k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes".ListerRegistry, chan struct {})
core/autoscaler_test.go:107: not enough arguments in call to NewAutoscaler
	have (AutoscalerOptions, *simulator.PredicateChecker, *"k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/client-go/kubernetes/fake".Clientset, record.EventRecorder, "k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes".ListerRegistry)
	want (AutoscalerOptions, *simulator.PredicateChecker, "k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/client-go/kubernetes".Interface, record.EventRecorder, "k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes".ListerRegistry, chan struct {})
```

File: cluster-autoscaler/cloudprovider/builder/cloud_provider_builder.go [hunk 1]
```diff
@@ -42,7 +48,7 @@ func NewCloudProviderBuilder(cloudProviderFlag string, cloudConfig string) Cloud
 }
 
 // Build a cloud provider from static settings contained in the builder and dynamic settings passed via args
-func (b CloudProviderBuilder) Build(discoveryOpts cloudprovider.NodeGroupDiscoveryOptions) cloudprovider.CloudProvider {
+func (b CloudProviderBuilder) Build(discoveryOpts cloudprovider.NodeGroupDiscoveryOptions, providerStopChannel chan struct{}) cloudprovider.CloudProvider {
```
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/builder/cloud_provider_builder.go]: break the line

Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/builder/cloud_provider_builder.go]: where do we actually use this stop channel.

File: cluster-autoscaler/cloudprovider/kubemark/kubemark_other.go [hunk 2]
```diff
@@ -1,3 +1,5 @@
+// +build ignore
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/kubemark/kubemark_other.go]: shouldn't that be +build !linux ? Also I'm pretty sure it needs to go below boilerplate

File: cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go [hunk 3]
```diff
@@ -1,3 +1,5 @@
+// +build linux
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go]: below boilerplate

File: cluster-autoscaler/cloudprovider/kubemark/kubemark_other.go [hunk 4]
```diff
@@ -14,8 +16,58 @@ See the License for the specific language governing permissions and
...
+
+func (kubemark *KubemarkCloudProvider) NewNodeGroup(name string, machineType string, labels map[string]string, extraResources map[string]resource.Quantity) (cloudprovider.NodeGroup, error) {
+	return nil, cloudprovider.ErrNotImplemented
+}
+
+type FakeNodeGroup struct{}
+
+func (f *FakeNodeGroup) Id() string                            { return "" }
+func (f *FakeNodeGroup) MinSize() int                          { return 1 }
+func (f *FakeNodeGroup) MaxSize() int                          { return 2 }
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/kubemark/kubemark_other.go]: Can we make MinSize and MaxSize 0? It's int zero value (like "" for string, false for bool or nil for ptr) and it feels more correct than arbitrary values.

File: cluster-autoscaler/core/autoscaler.go [hunk 5]
```diff
@@ -50,9 +50,9 @@ type Autoscaler interface {
 
 // NewAutoscaler creates an autoscaler of an appropriate type according to the parameters
 func NewAutoscaler(opts AutoscalerOptions, predicateChecker *simulator.PredicateChecker, kubeClient kube_client.Interface,
-	kubeEventRecorder kube_record.EventRecorder, listerRegistry kube_util.ListerRegistry) (Autoscaler, errors.AutoscalerError) {
+	kubeEventRecorder kube_record.EventRecorder, listerRegistry kube_util.ListerRegistry, providerStopChannel chan struct{}) (Autoscaler, errors.AutoscalerError) {
 
-	autoscalerBuilder := NewAutoscalerBuilder(opts.AutoscalingOptions, predicateChecker, kubeClient, kubeEventRecorder, listerRegistry)
+	autoscalerBuilder := NewAutoscalerBuilder(opts.AutoscalingOptions, predicateChecker, kubeClient, kubeEventRecorder, listerRegistry, providerStopChannel)
```
Comment (MaciekPytel) [inline cluster-autoscaler/core/autoscaler.go]: this changes are just to push through the stop channel, which we don't seem to be using anyway

File: cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go [hunk 6]
```diff
@@ -14,16 +16,243 @@ See the License for the specific language governing permissions and
...
+func (kubemark *KubemarkCloudProvider) Name() string {
+	return ProviderName
+}
+
+// NodeGroups returns all node groups configured for this cloud provider.
+func (kubemark *KubemarkCloudProvider) NodeGroups() []cloudprovider.NodeGroup {
+	result := make([]cloudprovider.NodeGroup, 0, len(kubemark.nodeGroups))
+	for _, nodegroup := range kubemark.nodeGroups {
+		result = append(result, nodegroup)
+	}
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go]: golang trivia - you can avoid the loop by doing `result = append(result, kubemark.nodeGroups...)`

Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go]: (not saying you should necessarily do that, I'm fine either way)

File: cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go [hunk 7]
```diff
@@ -14,16 +16,243 @@ See the License for the specific language governing permissions and
...
+
+// NodeGroupForNode returns the node group for the given node.
+func (kubemark *KubemarkCloudProvider) NodeGroupForNode(node *apiv1.Node) (cloudprovider.NodeGroup, error) {
+	nodeGroupName, err := kubemark.kubemarkController.GetNodeGroupForNode(node.ObjectMeta.Name)
+	if err != nil {
+		return nil, err
+	}
+	for i, nodeGroup := range kubemark.nodeGroups {
+		if nodeGroup.Name == nodeGroupName {
+			return kubemark.nodeGroups[i], nil
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go]: sorry, this is purely aesthetics but why not `return nodeGroup, nil`?

File: cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go [hunk 8]
```diff
@@ -14,16 +16,243 @@ See the License for the specific language governing permissions and
...
+	return nil, cloudprovider.ErrNotImplemented
+}
+
+// NodeGroup implements NodeGroup interfrace.
+type NodeGroup struct {
+	Name string
+
+	kubemarkController *kubemark.KubemarkController
+
+	minSize int
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go]: why linebreaks in this struct definition?

File: cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go (hunk 3 above)
Comment (bskiba) [inline cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go]: As discussed offline, can't be moved, only one line comments can go before the build directive.

File: cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go (hunk 8 above)
Comment (bskiba) [inline cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go]: Done

File: cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go (hunk 7 above)
Comment (bskiba) [inline cluster-autoscaler/cloudprovider/kubemark/kubemark_linux.go]: Done

File: cluster-autoscaler/cloudprovider/kubemark/kubemark_other.go (hunk 2 above)
Comment (bskiba) [inline cluster-autoscaler/cloudprovider/kubemark/kubemark_other.go]: Done, can't move due to boiletplate being a muti-line comment

File: cluster-autoscaler/cloudprovider/kubemark/kubemark_other.go (hunk 4 above)
Comment (bskiba) [inline cluster-autoscaler/cloudprovider/kubemark/kubemark_other.go]: Done

**Comment by bskiba** [comment]
Addressed comments, removed the channel that was unused.

**Comment by bskiba** [comment]
Squashed the commits.

**Pull request merged**